### ISSUE TITLE
HRQ: Resources are not removed from the quotas' usages when a namespace is moved out of a subtree

### ISSUE THREAD

**Issue by mzeevi** [kubernetes-sigs/hierarchical-namespaces#214]
Hi,

The HRQ [documentation](https://cloud.google.com/anthos-config-management/docs/how-to/using-hierarchical-resource-quotas) notes the following:

> When a namespace is [moved out of a subtree](https://github.com/kubernetes-sigs/hierarchical-namespaces/blob/master/docs/user-guide/how-to.md#use-full) with hierarchical quotas in its ancestors, it is no longer subject to these quotas, and its resources are removed from the quotas' usages.

However, this does not seem to be the case here.

**Steps to reproduce:**

1. Clone HNC repo and install `kubectl-hns`
2. Set up Kind cluster with cert-manager:
```
$ kind create cluster
$ kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.8.0/cert-manager.yaml
```
3. Deploy on Kind cluster:
```
$ make deploy-hrq HNC_IMG="localhost/hnc-manager:v0.0.1" CONFIG=kind
```
4. Set up hierarchy using `kubectl-hns`:
```
$ kubectl create ns root-ns
$ kubectl-hns create subns-1 -n root-ns
$ kubectl-hns create subns-2 -n root-ns
$ kubectl-hns create subns-3 -n subns-1
```
5. The `root-ns` tree now looks like this:
```
root-ns
├── [s] subns-1
│   └── [s] subns-3
└── [s] subns-2

[s] indicates subnamespaces
```
6. Create `HierarchicalResourceQuota` object with limit of 10 on configmaps on `root-ns`.
```
$ cat <<EOF | kubectl apply -f -
apiVersion: hnc.x-k8s.io/v1alpha2
kind: HierarchicalResourceQuota
metadata:
  name: example-hrq
  namespace: root-ns
spec:
  hard:
    configmaps: "10"
EOF
```
7. The `status.used` property of `HierarchicalResourceQuota/example-hrq` shows `Configmaps: 4`, corresponding to the 4 `kube-root-ca.crt` configmaps - one for each namespace/subns:
```
$ kubectl describe HierarchicalResourceQuota example-hrq -n root-ns

Name:         example-hrq
Namespace:    root-ns
Labels:       <none>
Annotations:  <none>
API Version:  hnc.x-k8s.io/v1alpha2
Kind:         HierarchicalResourceQuota
Metadata:
  Creation Timestamp:  2022-06-28T11:14:11Z
  Generation:          4
  Managed Fields:
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:hard:
          .:
          f:configmaps:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2022-06-28T11:14:11Z
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:hard:
          .:
          f:configmaps:
        f:used:
          .:
          f:configmaps:
    Manager:         manager
    Operation:       Update
    Time:            2022-06-28T11:14:12Z
  Resource Version:  4434
  UID:               a99f5217-f78a-45ab-8856-9547df97fb9b
Spec:
  Hard:
    Configmaps:  10
Status:
  Hard:
    Configmaps:  10
  Used:
    Configmaps:  4
Events:          <none>
```
8. Create a new full namespace and a config map in this namespace:
```
$ kubectl create ns test-ns
$ cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: test-cm  
  namespace: test-ns    
data:
 number: "3"
EOF
```
9. Move `test-ns` into `root-ns`:
```
$ kubectl-hns set test-ns --parent root-ns
```
10. The `status.used` property of `HierarchicalResourceQuota/example-hrq` shows `Configmaps: 6`, which makes sense and satisfies the following from the HRQ [documentation](https://cloud.google.com/anthos-config-management/docs/how-to/using-hierarchical-resource-quotas:

> When a namespace is added to a subtree with hierarchical quotas, it is subject to the hierarchical quotas, and its resource usages are added to the quotas' usages.

```
$ kubectl describe HierarchicalResourceQuota example-hrq -n root-ns

Name:         example-hrq
Namespace:    root-ns
Labels:       <none>
Annotations:  <none>
API Version:  hnc.x-k8s.io/v1alpha2
Kind:         HierarchicalResourceQuota
Metadata:
  Creation Timestamp:  2022-06-28T11:14:11Z
  Generation:          5
  Managed Fields:
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:hard:
          .:
          f:configmaps:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2022-06-28T11:14:11Z
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:hard:
          .:
          f:configmaps:
        f:used:
          .:
          f:configmaps:
    Manager:         manager
    Operation:       Update
    Time:            2022-06-28T11:14:12Z
  Resource Version:  4594
  UID:               a99f5217-f78a-45ab-8856-9547df97fb9b
Spec:
  Hard:
    Configmaps:  10
Status:
  Hard:
    Configmaps:  10
  Used:
    Configmaps:  6
Events:          <none>
```
10. Now move `test-ns` out of the `root-ns` subtree by setting it as root:
```
$ kubectl-hns set --root test-ns
```
11. The `status.used` property of `HierarchicalResourceQuota/example-hrq` still shows `Configmaps: 6`, but should show `Configmaps: 4`.

```
$ kubectl describe HierarchicalResourceQuota/example-hrq -n root-ns
Name:         example-hrq
Namespace:    root-ns
Labels:       <none>
Annotations:  <none>
API Version:  hnc.x-k8s.io/v1alpha2
Kind:         HierarchicalResourceQuota
Metadata:
  Creation Timestamp:  2022-06-28T11:14:11Z
  Generation:          5
  Managed Fields:
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:hard:
          .:
          f:configmaps:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2022-06-28T11:14:11Z
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:hard:
          .:
          f:configmaps:
        f:used:
          .:
          f:configmaps:
    Manager:         manager
    Operation:       Update
    Time:            2022-06-28T11:14:12Z
  Resource Version:  4594
  UID:               a99f5217-f78a-45ab-8856-9547df97fb9b
Spec:
  Hard:
    Configmaps:  10
Status:
  Hard:
    Configmaps:  10
  Used:
    Configmaps:  6
Events:          <none>
```

I'm aware that this feature is not yet production-ready, but is it a bug or something that hasn't been implemented yet or perhaps my usage here is wrong?

**Comment by mzeevi** [comment]
I tried to reproduce it today and noticed that after the controller is restarted (I deleted the pod and let the deployment create a new one), the status is shown correctly - 4 config maps in the status.

The update does not seem to be triggered without a restart for some reason

```
Name:         example-hrq
Namespace:    root-ns
Labels:       <none>
Annotations:  <none>
API Version:  hnc.x-k8s.io/v1alpha2
Kind:         HierarchicalResourceQuota
Metadata:
  Creation Timestamp:  2022-07-25T14:22:08Z
  Generation:          7
  Managed Fields:
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:hard:
          .:
          f:configmaps:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2022-07-25T14:22:08Z
    API Version:  hnc.x-k8s.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:hard:
          .:
          f:configmaps:
        f:used:
          .:
          f:configmaps:
    Manager:         manager
    Operation:       Update
    Time:            2022-07-25T14:22:08Z
  Resource Version:  1405
  UID:               3d5432fe-4b0c-4cd2-9356-d335328bf750
Spec:
  Hard:
    Configmaps:  10
Status:
  Hard:
    Configmaps:  10
  Used:
    Configmaps:  4
Events:          <none>
```

**Comment by adrianludwin** [comment]
Thanks for this! I'll try to look into it later this month but I'm focused
on a bunch of other stuff right now. Is this urgent for you? Are you able
to contribute patches yourself (e.g. write Go code, get Github PRs
reviewed, etc)? That could accelerate this process.

**Comment by mzeevi** [comment]
It's not urgent, no. 

I'll gladly take a look into the code and see if I can get it fixed. 

Any suggestions on where you think the problem might be?

**Comment by mzeevi** [comment]
I believe this should be fixed by #219

**Comment by adrianludwin** [comment]
Thanks, I've left comments in the PR. We shouldn't rely on the periodic sync, that's basically just to recover from _bugs_ and shouldn't be relied on in normal operation.

**Comment by adrianludwin** [comment]
Oh wait, this was fixed by #219 . Thanks!
/close
